Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/MarkerDetection/MarkerVisualizerCoordinateService.cs

```diff
@@ -48,56 +47,60 @@ public SpatialCoordinate(int id)
+            }
+
+            public void HideMarker()
+            {
+                locatedState = LocatedState.Resolved;
+                markerVisual.HideMarker();
+            }
         }
 
         private readonly IMarkerVisual markerVisual;
         private readonly UnityEngine.Transform markerInWorldSpace;
-        private readonly Func<int> generateMarkerId;
 
         private SpatialCoordinate markerCoordinate;
 
+        protected override bool SupportsDiscovery => false;
+
         public MarkerVisualizerCoordinateService(IMarkerVisual markerVisual, UnityEngine.Transform markerInWorldSpace, Func<int> generateMarkerId)
         {
             this.markerVisual = markerVisual ?? throw new ArgumentNullException(nameof(generateMarkerId));
             this.markerInWorldSpace = markerInWorldSpace;
-            this.generateMarkerId = generateMarkerId ?? throw new ArgumentNullException(nameof(generateMarkerId));
         }
 
         protected override void OnManagedDispose()
         {
             base.OnManagedDispose();
-
-            IsTracking = false;
         }
 
-        protected override async Task RunTrackingAsync(CancellationToken cancellationToken)
+        protected override bool TryParse(string id, out int result) => int.TryParse(id, out result);
+
+        protected override async Task OnDiscoverCoordinatesAsync(CancellationToken cancellationToken, int[] idsToLocate)
         {
-            ThrowIfDisposed();
+            if (idsToLocate == null || idsToLocate.Length < 1)
+            {
+                throw new ArgumentNullException($"{nameof(MarkerVisualizerCoordinateService)} depends on ids so that it could visualize them, at least one should be provided.");
+            }
 
-            markerCoordinate = new SpatialCoordinate(generateMarkerId()) { IsLocated = true };
+            markerCoordinate = new SpatialCoordinate(idsToLocate[0], markerVisual);
             OnNewCoordinate(markerCoordinate.Id, markerCoordinate);
 
-            markerVisual.ShowMarker(markerCoordinate.Id);
-
-            // TODO we can probably get rid of UpdateTick and use a Unity synchronization context here to wait frames
-            await Task.Delay(-1, cancellationToken).IgnoreCancellation();
-
-            markerVisual.HideMarker();
-            markerCoordinate.IsLocated = false;
-            markerCoordinate = null;
-        }
-
-        /// <summary>
-        /// Call this method each frame to process the marker position/rotation update.
-        /// </summary>
-        public void UpdateTick()
-        {
-            ThrowIfDisposed();
+            markerCoordinate.ShowMarker();
 
-            if (markerCoordinate != null)
+            while (cancellationToken.IsCancellationRequested)
             {
                 markerCoordinate.WorldToCoordinate = markerInWorldSpace.worldToLocalMatrix;
+                await Task.Delay(1, cancellationToken).IgnoreCancellation(); // Wait a frame
```

Comment:
This won't wait for a frame, it will wait for 1ms, correct? Or will it wait for a full frame because of the synchronization context of being on Unity's main thread? If that's the case, probably worth clarification about why in a comment here.

Reply:
Yeah, it will wait a frame because of Unity's synchronization context.

---
In reply to: [284813578](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4390#discussion_r284813578) [](ancestors = 284813578)

Reply:
I will add a comment.

---
In reply to: [284814467](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4390#discussion_r284814467) [](ancestors = 284814467,284813578)